src/uct/ib/base/ib_iface.c

```diff
@@ -1173,6 +1174,95 @@ static void uct_ib_iface_set_path_mtu(uct_ib_iface_t *iface,
     }
 }
 
+const char*
+uct_ib_iface_sl_mask_str(uint16_t sl_mask, ucs_string_buffer_t *str_b)
+{
+    uint8_t sl;
+
+    if (sl_mask == 0) {
+        ucs_string_buffer_appendf(str_b, "<none>");
+        goto out;
+    }
+
+    ucs_for_each_bit(sl, sl_mask) {
+        ucs_assert(sl < 16);
+
+        if (ucs_string_buffer_length(str_b) == 0) {
+            ucs_string_buffer_appendf(str_b, "%u", sl);
+        } else {
+            ucs_string_buffer_appendf(str_b, ", %u", sl);
+        }
+    }
+
+out:
+    return ucs_string_buffer_cstr(str_b);
+}
+
+
+ucs_status_t
+uct_ib_iface_select_sl(const uct_ib_iface_config_t *ib_config,
+                       ucs_ternary_value_t ar_enable, uint16_t ooo_sl_mask,
+                       const char *dev_port_name, uint8_t *sl_p)
+{
+    ucs_string_buffer_t sl_str;
+    uint8_t sl;
+
+    ucs_string_buffer_init(&sl_str);
+
+    /* Select the default  */
+    ucs_assert((ib_config->sl <= UINT8_MAX) || (ib_config->sl == UCS_ULUNITS_AUTO));
+    sl = (ib_config->sl == UCS_ULUNITS_AUTO) ? 0 : (uint8_t)ib_config->sl;
+
+    if (ar_enable == UCS_NO) {
+        goto out;
+    }
+
+    if (ib_config->sl == UCS_ULUNITS_AUTO) {
+        if (ooo_sl_mask != 0) {    
+            /* Select the first available SL with AR support, since a specific
+             * SL was not requested by user */
+           sl = ucs_ffs64_safe(ooo_sl_mask);
+           ucs_assert(sl != 64);
+        } else {
+            /* No SLs support AR */
+            if (ar_enable == UCS_YES) {
+                ucs_error("AR support is requested, but no SLs support AR on %s",
```

Comment:
btw, maybe worth changing this error message. If you can't detect AR SL it does not mean it is not supported.
Need to warn, like "AR support is requested, but could not find proper SL which supports it. Please set it manually"